### PR TITLE
Don't use intermediate file when listing contents of tar.xz file

### LINKED ISSUE #41190: Cannot extract tar.xz archive when it exceeds size of /tmp
### Description of Issue/Question

When using the archive.extracted state to extract a 10GB .tar.xz file, I get this error: "xz: (stdout): Write error: No space left on device".

I investigated the situation and found that in salt/modules/archive.py, in the _list_tar function, a temporary file is generated with tempfile.mkstemp and the compressed archive is then decompressed to that location. If the size of the decompressed .tar file exceeds the size of the /tmp filesystem, xz terminates with the error above.

I'm wondering why a temp file is written at all and xz output is not simply piped to tar.

### Versions Report
Salt Version:
           Salt: 2016.11.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.24.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.13 (default, Feb 11 2017, 12:22:40)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.1
            ZMQ: 4.2.1

System Versions:
           dist:
        machine: x86_64
        release: 4.10.13-1-ARCH
         system: Linux
        version: Not Installed

### PR DESCRIPTION
**Pull request by terminalmage**
This instead pipes the decompressed tar data directly to ``tarfile.open()``

Thanks to @dwfreed for working with me on this at PyCon!

Resolves #41190.

### CODE REVIEW

File: salt/modules/archive.py [hunk 1]
```diff
@@ -157,7 +158,10 @@ def _list_tar(name, cached, decompress_cmd, failhard=False):
         files = []
         links = []
         try:
-            with contextlib.closing(tarfile.open(cached)) as tar_archive:
+            open_kwargs = {'name': cached} \
+                if not isinstance(cached, subprocess.Popen) \
+                else {'fileobj': cached.stdout, 'mode': 'r|*'}
```
Comment (dwfreed) [inline salt/modules/archive.py]: You can actually drop the * here.  The subprocess should be returning an uncompressed tar archive, so there's no need to run the decompression heuristic on it.

Comment (terminalmage) [inline salt/modules/archive.py]: Good call, Just pushed another commit to drop the asterisk.

File: salt/modules/archive.py [hunk 2]
```diff
@@ -168,7 +172,15 @@ def _list_tar(name, cached, decompress_cmd, failhard=False):
             return dirs, files, links
 
         except tarfile.ReadError:
-            if not failhard:
+            if failhard:
+                if fileobj:
```
Comment (dwfreed) [inline salt/modules/archive.py]: This should be `if isinstance(cached, subprocess.Popen):`

Comment (terminalmage) [inline salt/modules/archive.py]: Yeah, that was leftover from earlier. I had initially added a ``fileobj`` arg to ``_list_tar()`` and then decided to remove it and instead just check if ``cached`` was a Popen object. I just didn't catch this other reference. Thanks!

**Review by cachedout — APPROVED** [review]
Clever! I dig it.

**Pull request merged**